extensions/mssql/src/credentialstore/nativeCredentialService.ts

```diff
@@ -0,0 +1,144 @@
+
+import { AppContext } from '../appContext';
+import { SqlOpsDataClient, ISqlOpsFeature, SqlOpsFeature } from 'dataprotocol-client';
+import * as Utils from '../utils';
+import { ClientCapabilities, RPCMessageType, ServerCapabilities } from 'vscode-languageclient';
+import * as Contracts from './contracts';
+import { Disposable, SecretStorage } from 'vscode';
+import * as azdata from 'azdata';
+import * as UUID from 'vscode-languageclient/lib/utils/uuid';
+
+export class NativeCredentialService extends SqlOpsFeature<any> {
+
+	private static readonly messagesTypes: RPCMessageType[] = [
+		Contracts.DeleteCredentialRequest.type,
+		Contracts.SaveCredentialRequest.type,
+		Contracts.ReadCredentialRequest.type
+	];
+
+	public static asFeature(context: AppContext): ISqlOpsFeature {
+		return class extends NativeCredentialService {
+			private _secretStorage: SecretStorage;
+			private _useNativeCredentialService: boolean;
+			private _passwordsMigrated: boolean;
+
+			constructor(client: SqlOpsDataClient) {
+				super(context, client);
+				this._secretStorage = context.extensionContext.secrets;
+				this._passwordsMigrated = context.extensionContext.globalState.get(Utils.configPasswordsMigrated);
+				this._useNativeCredentialService = Utils.useNativeCredentialsEnabled();
+			}
+
+			override fillClientCapabilities(capabilities: ClientCapabilities): void {
+				Utils.ensure(Utils.ensure(capabilities, 'credentials')!, 'credentials')!.dynamicRegistration = true;
+			}
+
+			override async initialize(capabilities: ServerCapabilities): Promise<void> {
+				this.register(this.messages, {
+					id: UUID.generateUuid(),
+					registerOptions: undefined
+				});
+				if (this._useNativeCredentialService && !this._passwordsMigrated) {
+					await this.migratePasswords();
+					this._passwordsMigrated = true;
+					await context.extensionContext.globalState.update(Utils.configPasswordsMigrated, this._passwordsMigrated);
+				}
+			}
+
+			/**
+			 * Removes a credential for a given connection profile from sqltoolsservice
+			 * and adds it to secret storage
+			 */
+			private async cleanCredential(conn: azdata.connection.ConnectionProfile): Promise<boolean> {
+				const credentialId = Utils.formatCredentialId(conn);
+				// read password from every saved password connection
+				const credential = await this._client.sendRequest(Contracts.ReadCredentialRequest.type, { credentialId, password: undefined });
+				if (credential.password) {
+					const password = credential.password;
+					// save it in secret store
+					await this._secretStorage.store(credentialId, password);
+					// check if it's saved
+					const savedPassword = await this._secretStorage.get(credentialId);
+					if (savedPassword === password) {
+						// delete from tools service
+						const result = await this._client.sendRequest(Contracts.DeleteCredentialRequest.type,
+							{ credentialId, password: undefined });
+						return result;
+					} else {
+						return false;
```

Comment:
Why are you bothering to return false? It's never even used so at best it's just misleading, but also it seems like we should be showing them an error and letting them know which ones failed to migrate if that happens